### DISCUSSION TITLE
Custom IOA - Send multiple field_values

### DISCUSSION THREAD

**Discussion by oscardecr** [CrowdStrike/falconpy#915] [Q&A] [answered]
Hi,

Is there a way to Send multiple field_values to create_rule method?

Let's say I have a rule with multiple field_values, something like:

    "field_values": [
        {
            "name": "name_1",
            "value": ".*",
            "label": "label_1",
            "type": "excludable",
            "values": [
                {
                    "label": "label_1",
                    "value": "value_1"
                }
            ],
            "final_value": "value_1"
        },
        {
            "name": "name_2",
            "value": ".*",
            "label": "label_2",
            "type": "excludable",
            "values": [
                {
                    "label": "label_2",
                    "value": "value_2*"
                }
            ],
            "final_value": "value_2"
        }

Documentation reference: https://www.falconpy.io/Service-Collections/Custom-IOA.html#create_rule

**Comment by jshcodes** [comment, marked as answer]
Hi @oscardecr -

In the current version ([v1.2.11](https://github.com/CrowdStrike/falconpy/releases/tag/v1.2.11)), this operation's `field_values` keyword will only accept one dictionary of values.  Expanding the payload handler to handle multiple dictionaries here is a solid enhancement ask, I'll get an issue created for it.

In the interim, you __can__ accomplish this by swapping over to leveraging the `body` keyword instead.

```python
from falconpy import CustomIOA

ioa = CustomIOA(client_id=CLIENT_ID, client_secret=CLIENT_SECRET)

body_payload = {
        "comment": "string",
        "description": "string",
        "name": "string",
        "platform": "string",
        "enabled": true,
        "field_values": [
            {
                "final_value": "string",
                "label": "string",
                "name": "value1",
                "type": "string",
                "value": "string",
                "values": [
                    {
                        "label": "string",
                        "value": "string"
                    }
                ]
            },
            {
                "final_value": "string",
                "label": "string",
                "name": "value2",
                "type": "string",
                "value": "string",
                "values": [
                    {
                        "label": "string",
                        "value": "string"
                    }
                ]
            }
        ]
}

result = ioa.create_rule(body=body_payload)

print(result)
```

Thank you for the question!! 😄

**Reply by oscardecr** [reply]
Hey Joshua - thanks for the rapid response. Sounds good, just tried to do what you suggested with the payload, however for some reason I get 400 status code with "invalid update-rules request"
![image](https://user-images.githubusercontent.com/94642678/219690975-316e18fe-1ca2-4375-867b-30739643a40c.png)

Interesting thing is, the rule is being uploaded successfully.

This is my code:

```
    body_payload = {'comment': comment, 'ruletype_id': ruletype_id, 'name': name, 'pattern_severity': pattern_severity,
                    'field_values': field_values, 'disposition_id': disposition_id, 'rulegroup_id': rulegroup_id}
    response = API_CustomIOA.create_rule(body=body_payload)
    status_code = response['status_code']

    if status_code == 200 or status_code == 201:
        pass
    else:
        raise ValueError(
            'Error pushing rule to Crowdstrike. Refer to the error message from Crowdstrike below..\n {}'.format(
                response))
```

**Reply by jshcodes** [reply]
That's weird... let me try to recreate this.

**Reply by oscardecr** [reply]
If this helps - I just rollback to an old version of falconpy I have installed, not sure what version, I believe it is V4. Just tried to push a rule with that version, using the same code I just shared with you, and it worked.

This seems to be the create_rule method for that old version I'm referring to:

    def create_rule(self: object, body: dict, cs_username: str) -> dict:
        """Create a rule within a rule group. Returns the rule."""
        # [POST] https://assets.falcon.crowdstrike.com/support/api/swagger.html#/custom-ioa/create-rule
        FULL_URL = self.base_url+'/ioarules/entities/rules/v1'
        HEADERS = self.headers
        HEADERS['X-CS-USERNAME'] = cs_username
        BODY = body
        returned = service_request(caller=self,
                                   method="POST",
                                   endpoint=FULL_URL,
                                   body=BODY,
                                   headers=HEADERS,
                                   verify=self.ssl_verify
                                   )
        return returned

**Reply by jshcodes** [reply]
What version were you on before?

**Reply by oscardecr** [reply]
@jshcodes on crowdstrike-falconpy 1.2.10

**Reply by jshcodes** [reply]
Can I see your entire `body` payload? (Please redact any specific IDs, keys, hostnames, etc.) I just need to see the fields you are providing and their data types. 

Variations of the following work without error for me, so I'm wondering if it's related to the specific request.

```python
import os
from falconpy import CustomIOA


ioa = CustomIOA(client_id=CLIENT_ID,
                client_secret=CLIENT_SECRET
                )

body_payload = {
        "comment": "FalconPy testing",
        "description": "Github question 915",
        "name": "question915",
        "platform": "linux",
        "enabled": False,
        "field_values": [
            {
                "final_value": "nc.*",
                "label": "Parent Image Filename",
                "name": "ParentImageFilename",
                "type": "excludable",
                "value": "nc.*",
                "values": [
                    {
                        "label": "include",
                        "value": "nc.*"
                    }
                ]
            },
            {
                "final_value": "bob.*",
                "label": "Image Filename",
                "name": "ImageFilename",
                "type": "excludable",
                "value": "bob.*",
                "values": [
                    {
                        "label": "include",
                        "value": "bob.*"
                    }
                ]
            }
        ],
        "rulegroup_id": "REDACTED",
        "ruletype_id": "11",
        "pattern_severity": "medium",
        "disposition_id": 20
}

result = ioa.create_rule(body=body_payload)
print(result)
```

**Reply by oscardecr** [reply]
I noticed your payload is basically the entire rule body. I was just sending "field_values". Let me try to do it that way and get back to you

**Reply by oscardecr** [reply]
Ok now after the change I get:
` 'errors': [{'code': 500, 'message': "Internal Server Error: Please provide trace-id='5c1907eb-5c54-4994-9abf-da4491e03d13' to support"}]}}
`

I'll get a payload example to provide

**Reply by oscardecr** [reply]
One reason could be that json format does not allow single quotes, and when I load the json from the python code, it transforms double quotes to single quotes. I'll change that and test.

**Reply by jshcodes** [reply]
Did you specify a `rulegroup_id` in your payload?

**Reply by oscardecr** [reply]
Yes

**Reply by oscardecr** [reply]
I haven't found a way to change the dictionary single quotes to double quotes. That's the way json.load() works. I'll see if I can get a workaround. Although not sure this causes the issue.

**Reply by oscardecr** [reply]
Example payload:
{
    "created_by": "oscar@test.com",
    "created_on": "2021-03-08T13:24:25.329735886Z",
    "ruletype_id": "1",
    "ruletype_name": "Process Creation",
    "comment": "",
    "enabled": true,
    "name": "GIS_TEST_RULE_RODRIO",
    "description": "Test",
    "pattern_severity": "low",
    "action_label": "Detect",
    "disposition_id": 20,
    "field_values": [
        {
            "name": "GrandparentImageFilename",
            "value": ".*",
            "label": "Grandparent Image Filename",
            "type": "excludable",
            "values": [
                {
                    "label": "include",
                    "value": ".*"
                }
            ],
            "final_value": ".*"
        },
        {
            "name": "GrandparentCommandLine",
            "value": ".*",
            "label": "Grandparent Command Line",
            "type": "excludable",
            "values": [
                {
                    "label": "include",
                    "value": ".*"
                }
            ],
            "final_value": ".*"
        },
        {
            "name": "ParentImageFilename",
            "value": ".*",
            "label": "Parent Image Filename",
            "type": "excludable",
            "values": [
                {
                    "label": "include",
                    "value": ".*"
                }
            ],
            "final_value": ".*"
        },
        {
            "name": "ParentCommandLine",
            "value": ".*",
            "label": "Parent Command Line",
            "type": "excludable",
            "values": [
                {
                    "label": "include",
                    "value": ".*"
                }
            ],
            "final_value": ".*"
        },
        {
            "name": "ImageFilename",
            "value": ".*",
            "label": "Image Filename",
            "type": "excludable",
            "values": [
                {
                    "label": "include",
                    "value": ".*"
                }
            ],
            "final_value": ".*"
        },
        {
            "name": "CommandLine",
            "value": "(?i)wmic.+process.+lsass.+Processid.+",
            "label": "Command Line",
            "type": "excludable",
            "values": [
                {
                    "label": "include",
                    "value": "wmic.+process.+lsass.+Processid.+"
                }
            ],
            "final_value": "(?i)wmic.+process.+lsass.+Processid.+"
        }
    ]
}

**Reply by jshcodes** [reply]
There's no "rulegroup_id" value in the above.  Once I added that and gave it a valid Rule Group ID I was able to submit this payload successfully.

Here's the entire code I executed:

```python
import os
from falconpy import CustomIOA


ioa = CustomIOA(client_id=CLIENT_ID,
                client_secret=CLIENT_SECRET
                )

body_payload = {
	"created_by": "oscar@test.com",
	"created_on": "2021-03-08T13:24:25.329735886Z",
	"ruletype_id": "1",
        "rulegroup_id": "RULEGROUP ID HERE",
	"ruletype_name": "Process Creation",
	"comment": "",
	"enabled": True,
	"name": "GIS_TEST_RULE_RODRIO",
	"description": "Test",
	"pattern_severity": "low",
	"action_label": "Detect",
	"disposition_id": 20,
	"field_values": [{
			"name": "GrandparentImageFilename",
			"value": ".",
			"label": "Grandparent Image Filename",
			"type": "excludable",
			"values": [{
				"label": "include",
				"value": "."
			}],
			"final_value": "."
		},
		{
			"name": "GrandparentCommandLine",
			"value": ".",
			"label": "Grandparent Command Line",
			"type": "excludable",
			"values": [{
				"label": "include",
				"value": "."
			}],
			"final_value": "."
		},
		{
			"name": "ParentImageFilename",
			"value": ".",
			"label": "Parent Image Filename",
			"type": "excludable",
			"values": [{
				"label": "include",
				"value": "."
			}],
			"final_value": "."
		},
		{
			"name": "ParentCommandLine",
			"value": ".",
			"label": "Parent Command Line",
			"type": "excludable",
			"values": [{
				"label": "include",
				"value": "."
			}],
			"final_value": "."
		},
		{
			"name": "ImageFilename",
			"value": ".",
			"label": "Image Filename",
			"type": "excludable",
			"values": [{
				"label": "include",
				"value": "."
			}],
			"final_value": ".*"
		},
		{
			"name": "CommandLine",
			"value": "(?i)wmic.+process.+lsass.+Processid.+",
			"label": "Command Line",
			"type": "excludable",
			"values": [{
				"label": "include",
				"value": "wmic.+process.+lsass.+Processid.+"
			}],
			"final_value": "(?i)wmic.+process.+lsass.+Processid.+"
		}
	]
}

result = ioa.create_rule(body=body_payload)
print(result)
```

##### Result

```json
{'status_code': 201, 'headers': {'Server': 'nginx', 'Date': 'Fri, 17 Feb 2023 21:07:16 GMT', 'Content-Type': 'application/json', 'Content-Length': '770', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains, max-age=31536000; includeSubDomains', 'X-Cs-Region': 'us-1', 'X-Cs-Traceid': '96bbb528-REDACTED-66a860d7cf97', 'X-Ratelimit-Limit': '6000', 'X-Ratelimit-Remaining': '5999'}, 'body': {'meta': {'query_time': 0.095628843, 'writes': {'resources_affected': 1}, 'powered_by': 'svc-ioarules', 'trace_id': '96bbb528-REDACTED-66a860d7cf97'}, 'resources': [{'instance_id': '30', 'customer_id': 'REDACTED', 'ruletype_id': '1', 'ruletype_name': 'Process Creation', 'comment': '', 'enabled': False, 'deleted': False, 'magic_cookie': 3, 'rulegroup_id': 'REDACTED', 'version_ids': ['1'], 'instance_version': 1, 'name': 'GIS_TEST_RULE_RODRIO', 'description': 'Test', 'pattern_id': '41003', 'pattern_severity': 'low', 'action_label': 'Detect', 'disposition_id': 20, 'field_values': [{'name': 'GrandparentImageFilename', 'value': '(?i).', 'label': 'Grandparent Image Filename', 'type': 'excludable', 'values': [{'label': 'include', 'value': '.'}], 'final_value': '(?i).'}, {'name': 'GrandparentCommandLine', 'value': '(?i).', 'label': 'Grandparent Command Line', 'type': 'excludable', 'values': [{'label': 'include', 'value': '.'}], 'final_value': '(?i).'}, {'name': 'ParentImageFilename', 'value': '(?i).', 'label': 'Parent Image Filename', 'type': 'excludable', 'values': [{'label': 'include', 'value': '.'}], 'final_value': '(?i).'}, {'name': 'ParentCommandLine', 'value': '(?i).', 'label': 'Parent Command Line', 'type': 'excludable', 'values': [{'label': 'include', 'value': '.'}], 'final_value': '(?i).'}, {'name': 'ImageFilename', 'value': '(?i).', 'label': 'Image Filename', 'type': 'excludable', 'values': [{'label': 'include', 'value': '.'}], 'final_value': '(?i).'}, {'name': 'CommandLine', 'value': '(?i)wmic.+process.+lsass.+Processid.+', 'label': 'Command Line', 'type': 'excludable', 'values': [{'label': 'include', 'value': 'wmic.+process.+lsass.+Processid.+'}], 'final_value': '(?i)wmic.+process.+lsass.+Processid.+'}], 'created_by': 'REDACTED', 'created_on': '2023-02-17T21:07:16.810080207Z', 'modified_by': 'REDACTED', 'modified_on': '2023-02-17T21:07:16.810080207Z', 'committed_on': '0001-01-01T00:00:00Z'}]}}
```

**Reply by oscardecr** [reply]
Ah - my apologies. For some reason I read ruletype_id. Just performed a test where I provided rulegroup_id, got same error:

'errors': [{'code': 400, 'message': 'invalid create-rule request'}]}}

**Reply by jshcodes** [reply]
Is it a valid Rule Group ID?  I did the payload above without issue, that should be the only difference between the two.

**Reply by oscardecr** [reply]
Yes it is. That leaves me with the single quotes doubt, which I need to resolve.

**Reply by jshcodes** [reply]
I'm not using json.loads at all.  Just providing the payload dictionary directly to the keyword.

**Reply by oscardecr** [reply]
Due to the nature of the program I can't skip the json.load(). But I guess I can find a workaround. For now, I'll provide the payload directly just to test and see if the json stuff is causing the issue.

**Reply by oscardecr** [reply]
Hey Joshua - providing the body_payload worked. The issue was on my code. Thanks for all the help :)
